core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala

```diff
@@ -798,14 +815,52 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
     stale.foreach { log =>
       if (log.appId.isEmpty) {
         logInfo(s"Deleting invalid / corrupt event log ${log.logPath}")
-        deleteLog(new Path(log.logPath))
+        deleteLog(fs, new Path(log.logPath))
         listing.delete(classOf[LogInfo], log.logPath)
       }
     }
     // Clean the blacklist from the expired entries.
     clearBlacklist(CLEAN_INTERVAL_S)
   }
 
+  /**
+   * Delete driver logs from the configured spark dfs dir that exceed the configured max age
+   */
+  private[history] def cleanDriverLogs(): Unit = Utils.tryLog {
+    val driverLogDir = conf.get(DRIVER_LOG_DFS_DIR).get
+    val currentTime = clock.getTimeMillis()
+    val maxTime = currentTime - conf.get(MAX_DRIVER_LOG_AGE_S) * 1000
+    val appDirs = driverLogFs.get.listLocatedStatus(new Path(driverLogDir))
+    while (appDirs.hasNext()) {
+      val appDirStatus = appDirs.next()
+      if (appDirStatus.isDirectory()) {
+        val logFiles = driverLogFs.get.listStatus(appDirStatus.getPath())
```

Comment:
Not sure how I missed this before, but why are you storing log files in a subdirectory? What's the advantage? You'll have a ton of directories with a single log file. Instead, just write the log files directly into the parent directory... that also makes this code simpler / faster (less RPCs).